### ISSUE TITLE
filemtime(): stat failed when changing file path in FileSystemWatcher

### ISSUE THREAD

**Issue by mark-zammit** [ElementaryFramework/FireFS#1]
PHP throws warnings "filemtime(): stat failed..." if a file no longer exists when being actioned with FileSystemWatcher.

Simple fix is to check if file_exists before updating the mod time cache and removing it from watch if it no longer exists.

`private function _cacheLastModTimes() 
 {
        foreach ($this->_filesCache as $name => $path) {
        	if(!file_exists($path)) {
        		$this->_removeFromWatch($path);
			} else {
				$this->_lastModTimeCache[$path] = $this->_lmt($path);
			}
        }
    }`

**Comment by na2axl** [comment]
Hi @iiMarco, sorry for the delay (it's not really easy to work with the outbreak 😞).

This is a strange catch, because the file system is always scanned for changes before caching mod-times, so `_cacheModTimes()` will, normally, always cache mod-times of existing files.

Can you give more details please ? I mean, the command used to start the file watcher, the error stacktrace, and when the error occurs (on starting the watcher, during watching, or when stopping it).

**Comment by na2axl** [comment]
I'm going to close this issue for no response.

**Comment by mark-zammit** [comment]
Hi @na2axl, sorry, I put this code down for a while didn't have examples to help troubleshoot it more.

Sequence of events:

- `build()`
- Extended `build()` to touch existing files before `start()` so that existing files trigger modified event
- Callback triggered with file path of modified file
- File is read and uploaded to an SFTP and then deleted (move operation)
- Exception thrown

Operating System: CentOS 7

Calling code:

_NOTE: I've extended the watcher to include `setWatchIterations()` and `watchExistingFiles()` which sets a hard limit on process() iterations and touches existing files before running process() the first time, respectively._

```
// Configure file watcher
$this->watcher
->setListener(new FileSystemListener([
	"before" => $event_before_callback,
	"after" => $event_after_callback,
	"any" => $event_any_callback,
	"created" => $event_created_callback,
	"modified" => $event_modified_callback,
	"deleted" => $event_deleted_callback,
]))
->setRecursive(false)
->setPath($path)
->addPattern($this->getFileMask())
->setWatchInterval($this->getInterval())
->setWatchIterations($this->getIterations())
->watchExistingFiles(true)
->build()
;
// Start listening
$this->watcher->start();
```

Stack trace (i've removed path references and replace with '...'):
```
PHP Fatal error:  Uncaught Exception: filemtime(): stat failed for /.../PLAY_PRODUCT_UL_060821155838.csv in /.../Watcher.php:244
Stack trace:
#0 [internal function]: Z\FileSystem\Watcher->Z\FileSystem\{closure}(2, 'filemtime(): st...', '...', 1016, Array)
#1 /.../vendor/elementaryframework/fire-fs/src/FireFS/FireFS.php(1016): filemtime('...')
#2 /.../vendor/elementaryframework/fire-fs/src/FireFS/Watcher/FileSystemWatcher.php(442): ElementaryFramework\FireFS\FireFS->lastModTime('/...')
#3 /.../vendor/elementaryframework/fire-fs/src/FireFS/Watcher/FileSystemWatcher.php(436): ElementaryFramework\FireFS\Watcher\FileSystemWatcher->_lmt('/...')
#4 /.../vendor/elementaryframework/fire-fs/src/FireFS/Watcher/FileSystemWatcher.php(324): ElementaryFramework\FireFS\Watcher\FileSystemWatch in /.../Watcher.php on line 244
```

The only way i've been able to silence the warnings is to add the code I referenced previously which does a sanity check to see if the file still exists before caching changes.

This is only a problem for me because the `_cacheLastModTimes()` method and other internal class methods are `private` and not `protected` otherwise I could simply extend the solution. Because they are `private` I need to edit the source code method after each update.

Hope this details helps!

**Comment by na2axl** [comment]
Hi @iiMarco, thanks for giving more details on this. With your explanation I think the issue is because you are deleting a watched file inside the callback, so after the scan process, the watcher for now is not notified of this change so the cached list of file paths is not updated.

Your fix will work but your code breaks the behavior because you are removing a file from the watcher before the scan process, so that file may will never be notified for deletion.

The simpler solution can be to just check if the file exists before to get the `lastmodtime` as you have done, but without deleting it from the watcher, and then let the next scan process do the notification + deletion.

**Comment by na2axl** [comment]
@iiMarco I've opened the PR #5 for this issue, please when you got time check if it solves your problem.

**Comment by mark-zammit** [comment]
Hi @na2axl, that has solved the issue and definitely the simpler solution. Thanks so much for getting this sorted so quickly.

**Comment by na2axl** [comment]
Thanks @iiMarco, the PR has been merged and a new version (v1.2.1) has been released.
